Add operational document on configuration scoping (#14702)
istio/istio.io: content/en/docs/ops/configuration/mesh/configuration-scoping/index.md

---
title: Configuration Scoping
description: Shows how to scope configuration in Istio, for operational and performance benefits.

> weight: 10

Reviewer: If you want this doc to be first in the section then you should make the weight <10, e.g., 5. If you want it last it should be >50. Otherwise you need to pick a number to place it where you want, relative to the other doc weights.
```suggestion
weight: 60
```

keywords: [scalability]
owner: istio/wg-networking-maintainers
test: no
---

In order to program the service mesh, the Istio control plane (Istiod) reads a variety of configurations, including core Kubernetes types like `Service` and `Node`,
and Istio's own types like `Gateway`.

> These are then sent the data plane (see [Architecture](/docs/ops/deployment/architecture/) for more information).

Reviewer: ```suggestion
These are then sent to the data plane (see [Architecture](/docs/ops/deployment/architecture/) for more information).
```

## Scoping mechanisms

Istio offers a few tools to help control the scope of a configuration to meet different use cases.
Depending on your requirements, these can be used alone or together.

* `Sidecar` provides a mechanism for specific workloads to _import_ a set of configurations
* `exportTo` provides a mechanism to _export_ a configuration to a set of workloads
* `discoverySelectors` provides a mechanism to let Istio completely ignore a set of configurations

### `Sidecar` import

The [`egress.hosts`](/docs/reference/config/networking/sidecar/#IstioEgressListener) field in `Sidecar`
allows specifying a list of configurations to import.
Only configurations matching the specified criteria will be seen by sidecars impacted by the `Sidecar` resource.

For example:

{{< text yaml >}}
apiVersion: networking.istio.io/v1beta1
kind: Sidecar
metadata:
  name: default
spec:
  egress:
  - hosts:
    - "./*" # Import all configuration from our own namespace
    - "bookinfo/*" # Import all configuration from the bookinfo namespace
    - "external-services/example.com" # Import only 'example.com' from the external-services namespace
{{< /text >}}

### `exportTo`

> Istio's `VirtualService`, `DestinationRule`, and `ServiceEntry` allow provide a `spec.exportTo` field.

Reviewer 2: ```suggestion
Istio's `VirtualService`, `DestinationRule`, and `ServiceEntry` provide a `spec.exportTo` field.
```

Similarly, `Service` can be configured with the `networking.istio.io/exportTo` annotation.

Unlike `Sidecar` which allows a workload owner to control what dependencies it has, `exportTo` works in the opposite way, and allows the service owners to control
their own service's visibility.

For example, this configuration makes the `details` `Service` only visible to its own namespace, and the `client` namespace:

{{< text yaml >}}
apiVersion: v1
kind: Service
metadata:
  name: details
  annotations:
    networking.istio.io/exportTo: ".,client"
spec: ...

> {{</ text >}}

Reviewer: I think this is the source of your lint errors

```suggestion
{{< /text >}}
```

### `DiscoverySelectors`

While the previous controls operate on a workload or service owner level, [`DiscoverySelectors`](/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig) provides mesh wide control over configuration visibility.
Discovery selectors allows specifying criteria for which namespaces should be visible to the control plane.
Any namespaces not matching are ignored by the control plane entirely.

This can be configured as part of `meshConfig` during installation. For example:

{{< text yaml >}}
meshConfig:
  discoverySelectors:
    - matchLabels:
        # Allow any namespaces with `istio-discovery=enabled`
        istio-discovery: enabled
    - matchLabels:
        # Allow "kube-system"; Kubernetes automatically adds this label to each namespace
        kubernetes.io/metadata.name: kube-system
{{</ text >}}

{{< warning >}}
Istiod will always open a watch to Kubernetes for all namespaces.
However, discovery selectors will ignore objects that are not selected very early in its processing, minimizing costs.
{{</ warning >}}

## Frequently asked questions

> ### How can I understand the cost of a certain configuration

Reviewer 3: ```suggestion
### How can I understand the cost of a certain configuration?
```

Some tools to detect which configurations are changing:
* Istiod will log each change like: `Push debounce stable 1 for config Gateway/default/gateway: ..., full=true`.
  This shows a `Gateway` object in the `default` namespace changed. `full=false` would represent and optimized update such as `Endpoint`.
  Note: changes to `Service` and `Endpoints` will all show as `ServiceEntry`.
* Istiod exposes metrics `pilot_k8s_cfg_events` and `pilot_k8s_reg_events` for each change.
* `kubectl get <resource> --watch -oyaml --show-managed-fields` can show changes to an object (or objects) to help understand what is changing, and by whom.

[Headless services](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) (besides ones declared as [HTTP](/docs/ops/configuration/traffic-management/protocol-selection/#explicit-protocol-selection))
scale with the number of instances. This makes large headless services expensive, and a good candidate for exclusion with `exportTo` or equivalent.

### What happens if I connect to a service outside of my scope?

> When connecting to a service that has been excluded through one of the scoping mechanisms, they data plane will not know anything about the destination,

Reviewer 2: ```suggestion
When connecting to a service that has been excluded through one of the scoping mechanisms, the data plane will not know anything about the destination,
```

so it will be treated as [Unmatched traffic](/docs/ops/configuration/traffic-management/traffic-routing/#unmatched-traffic).

### What about Gateways?

While [Gateways](/docs/setup/additional-setup/gateway/) will respect `exportTo` and `DiscoverySelectors`, `Sidecar` objects do not impact Gateways.
However, unlike sidecars, gateways do not have configuration for the entire cluster by default.
Instead, each configuration is explicitly attached to the gateway, which mostly avoids this problem.

However, [currently](https://github.com/istio/istio/issues/29131) part of the data plane configuration (a "cluster", in Envoy terms), is always sent for

> the entire cluster, even if its not referenced explicitly.

Reviewer 2: ```suggestion
the entire cluster, even if it is not referenced explicitly.
```
